### PR TITLE
fix(nvmx): reentrant controller event notification

### PR DESCRIPTION
**Pull request by mtzaurus**
Current implementation of NVMe controller events could lead to deadlocks, when event listener being notified indirectly triggers the second event notification action for the same NVMe controller, since NVme controller performs event notification under protection of a mutex.
This fix makes event invocation deadlock-free.

Signed-off-by: Mikhail Tcymbaliuk <mtzaurus@gmail.com>

### CODE REVIEW

**Comment by mtzaurus** [comment]
bors try

File: io-engine/src/core/device_events.rs [hunk 1]
```diff
@@ -135,19 +135,42 @@ impl DeviceEventDispatcher {
...
+        // Invoke all listeners once the mutex is dropped.
+        let notified = {
+            if !listeners.is_empty() {
+                for l in &listeners {
+                    l.dispatch_event(evt, dev_name);
+                }
+                listeners.len()
+            } else {
+                0
+            }
```
Comment (tiagolobocastro) [inline io-engine/src/core/device_events.rs]: nit: can skip the check?

```suggestion
            for l in &listeners {
                l.dispatch_event(evt, dev_name);
            }
            listeners.len()
```

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/core/device_events.rs (hunk 1 above)
Comment (mtzaurus) [reply io-engine/src/core/device_events.rs]: Yup, looks shorter. :-)

**Review by dsavitskiy — APPROVED** [review]

**Comment by mtzaurus** [comment]
bors try

**Comment by mtzaurus** [comment]
bors merge

**Pull request merged**